utils/gridlink_impl.c.src

```diff
@@ -91,86 +88,52 @@ cellarray_DOUBLE * gridlink_DOUBLE(const int64_t NPART,
 
     const int64_t totncells = (int64_t) nmesh_x * (int64_t) nmesh_y * (int64_t) nmesh_z;
 
-    const DOUBLE xdiff = (options->periodic && options->boxsize > 0) ? options->boxsize:xmax-xmin;
-    const DOUBLE ydiff = (options->periodic && options->boxsize > 0) ? options->boxsize:ymax-ymin;
-    const DOUBLE zdiff = (options->periodic && options->boxsize > 0) ? options->boxsize:zmax-zmin;
-
-    const DOUBLE cell_volume = xbinsize * ybinsize * zbinsize;
-    const DOUBLE box_volume = xdiff * ydiff * zdiff;
-    int64_t expected_n=(int64_t)(NPART * cell_volume/box_volume * MEMORY_INCREASE_FAC);
-    expected_n=expected_n < 2 ? 2:expected_n;
-
     if(options->verbose) {
       fprintf(stderr,"In %s> Running with [nmesh_x, nmesh_y, nmesh_z]  = %d,%d,%d. ",__FUNCTION__,nmesh_x,nmesh_y,nmesh_z);
     }
 
     cellarray_DOUBLE *lattice  = (cellarray_DOUBLE *) my_malloc(sizeof(*lattice), totncells);
-    int64_t *all_cell_indices = NULL;
+    DOUBLE *realX = NULL, *realY = NULL, *realZ = NULL;  // The actual allocations backing the lattice
+    DOUBLE *realW[MAX_NUM_WEIGHTS] = {NULL};  // This initializes the whole array to NULL
+    int w_alloc_status = 0;
+    int num_weights = (WEIGHTS == NULL) ? 0 : WEIGHTS->num_weights;
+    
+    int64_t *cellstarts = (int64_t *) my_malloc(sizeof(int64_t), totncells);
+    int64_t *all_cell_indices = (int64_t *) my_malloc(sizeof(*all_cell_indices), NPART);
     int64_t *original_indices =  NULL;
-    int64_t *nallocated = NULL;//to keep track of how many particles have been allocated per cell (only when creating a copy of particle positions)
     if(options->copy_particles) {
-        nallocated = (int64_t *) my_malloc(sizeof(*nallocated), totncells);
+        realX = (DOUBLE *) my_malloc(sizeof(DOUBLE), NPART);
+        realY = (DOUBLE *) my_malloc(sizeof(DOUBLE), NPART);
+        realZ = (DOUBLE *) my_malloc(sizeof(DOUBLE), NPART);
+        
+        for(int w = 0; w < num_weights; w++){
+            realW[w] = (DOUBLE *) my_malloc(sizeof(DOUBLE), NPART);
+            w_alloc_status |= realW[w] == NULL;
+        }
     } else {
-        all_cell_indices = (int64_t *) my_malloc(sizeof(*all_cell_indices), NPART);
         original_indices = (int64_t *) my_malloc(sizeof(*original_indices), NPART);
     }
-    if(lattice == NULL ||
-       (options->copy_particles == 0 && all_cell_indices == NULL) ||
+    if(lattice == NULL || all_cell_indices == NULL ||
        (options->copy_particles == 0 && original_indices == NULL) ||
-       (options->copy_particles && nallocated == NULL)) {
+       (options->copy_particles && (realX == NULL || realY == NULL || realZ == NULL || cellstarts == NULL || w_alloc_status != 0))){
 
-        free(lattice);free(nallocated);free(all_cell_indices);free(original_indices);
+        free(lattice);free(realX);free(realY);free(realZ);free(all_cell_indices);free(original_indices);
+        for(int w = 0; w < MAX_NUM_WEIGHTS; w++){
```

Comment:
Since the allocation is done up to `num_weights`, should the upper limit here be `num_weights`?

Reply:
Yes, that's a bug, thanks!